source/server-discovery-and-monitoring/tests/README.rst

```diff
@@ -134,3 +137,201 @@ the driver's current TopologyDescription or ServerDescription.
+    the new primary quickly.
+
+7.  Configure the server to hang up on all "find" commands (using the
+    "failCommand" failpoint). Execute a find command and assert the
+    client marks the server Unknown. (See "Network error when reading
+    or writing" in the main design doc.)
+
+8.  Test that a MongoClient ignores errors from previous generations.
+
+    #. Create a MongoClient with ``retryWrites=false``.
+
+    #. Run an insert operation to prove the client has discovered the server.
+
+    #. Configure the following failpoint to block insert command for 500ms
+       and then close the connection. Blocking the insert commands allows
+       both commands to be executed on connections from the same pool
+       generation::
+
+         db.adminCommand({
+             configureFailPoint: "failCommand",
+             mode: {times: 2},
+             data: {
+               failCommands: ["insert"],
+               closeConnection: true,
+               blockConnection: true,
+               blockTimeMS: 500,
+             },
+         });
+
+    #. Run 2 insertOne operations concurrently and assert that both operations
+       fail with network errors.
+
+    #. Assert that the server is reset to Unknown exactly once and the
+       application pool is cleared exactly once. For example, assert that
+       there was a single single ServerDescriptionChangedEvent that marks the server Unknown and a single PoolClearEvent.
+
+    #. Run an insert operation to prove the server is rediscovered.
+
+9.  Issue a write from 2 threads using two connections at the same time.
+    Cause the server to fail both operations with a State Change
+    Error (using the failCommand failpoint). Assert that the server
+    is only reset to Unknown once and the application pool is not
+    cleared.
+
+RTT Tests
+~~~~~~~~~
+
+Run the following test(s) on MongoDB 4.4+.
+
+1.  Test that RTT is continuously updated.
+
+    #. Create a client with  ``heartbeatFrequencyMS=500``,
+       ``appName=streamingRttTest``, and subscribe to server events.
+
+    #. Run a find command to wait for the server to be discovered.
+
+    #. Sleep for 2 seconds. This must be long enough for multiple heartbeats
+       to succeed.
+
+    #. Assert that each ``ServerDescriptionChangedEvent`` includes a non-zero
+       RTT.
+
+    #. Configure the following failpoint to block isMaster commands for 250ms
+       which should add extra latency to each RTT check::
+
+         db.adminCommand({
+             configureFailPoint: "failCommand",
+             mode: {times: 1000},
+             data: {
+               failCommands: ["isMaster"],
+               blockConnection: true,
+               blockTimeMS: 500,
+               appName: "streamingRttTest",
+             },
+         });
+
+    #. Wait for the server's RTT to exceed 250ms. Eventually the average RTT
+       should also exceed 500ms but we use 250ms to speed up the test.
```

Comment:
I've implemented this test in Python and discovered it might not be possible in all drivers. SDAM suppresses "equal" `ServerDescriptionChangedEvents` so this test only ever publishes a single event. My python implementation uses an internal method to grab the current RTT and assert that it exceeds >=250ms.

The only way to fix this I can think of is to remove `ServerDescriptionChangedEvent` equality ([SPEC-1176](https://jira.mongodb.org/browse/SPEC-1176)) and always publish events. What do you think?

Reply:
I don't recall: are `TopologyChangedEvent`s published?  If so, you can register a listener for those and then you don't need the helper.

Reply:
In python we suppress both the ServerDescriptionChangedEvent and TopologyChangedEvent. At least that's how I interpret [this line](https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#server-description-equality):

> For the purpose of determining whether to publish SDAM events, two server descriptions having the same address MUST be considered equal if and only if the values of ServerDescription fields marked (=) are respectively equal.

Reply:
I added a note to this test:
```
    #. Wait for the server's RTT to exceed 250ms. Eventually the average RTT
       should also exceed 500ms but we use 250ms to speed up the test. Note
       that the `Server Description Equality`_ means that
       ServerDescriptionChangedEvents will not be published. This test may
       need to use a driver specific helper to obtain the latest RTT instead.
```